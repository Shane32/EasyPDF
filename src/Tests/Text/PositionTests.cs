using System.Drawing;
using Font = Shane32.EasyPDF.Font;

namespace Tests.Text;

public class PositionTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public PositionTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Theory]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.18f, 0.03f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1f, 6f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.18f, 0f, 0f, 1f, 0f, 0.875f)]
    [InlineData(1.2f, 6f, 1.2f, 0.1324f, 0.1366f, 0.043f, 0.216f, 0.036f, 0.083f, 1f, 0f, 0.875f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 0f, 1.05f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1f, 1f, 1.027f)]
    [InlineData(1f, 0f, 1f, 0.1103f, 0.1138f, 0.036f, 0.15f, 0f, 0f, 1.2f, 1f, 1.233f)]
    public void TextSizes(float lineSpacing, float paragraphSpacing, float stretchY, float capHeight, float ascent, float descent, float height, float leading, float paragraphLeading, float stretchX, float characterSpacing, float width)
    {
        _writer.Font = new Font(StandardFonts.Times, 12f) {
            LineSpacing = lineSpacing,
            ParagraphSpacing = paragraphSpacing,
            StretchY = stretchY,
            CharacterSpacing = characterSpacing,
            StretchX = stretchX,
        };
        _writer.TextCapHeight().ShouldBe(capHeight, 0.001f);
        _writer.TextAscent().ShouldBe(ascent, 0.001f);
        _writer.TextDescent().ShouldBe(descent, 0.001f);
        _writer.TextHeight().ShouldBe(height, 0.001f);
        _writer.TextLeading().ShouldBe(leading, 0.001f);
        _writer.TextHeight().ShouldBe(_writer.TextAscent() + _writer.TextDescent() + _writer.TextLeading(), 0.001f);
        _writer.TextParagraphLeading().ShouldBe(paragraphLeading, 0.001f);
        _writer.TextWidth("Hello World!").ShouldBe(width, 0.001f);
    }
    
    [Fact]
    public void Standard()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedY_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f, LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void StretchedX()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { StretchX = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void CharacterSpacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { CharacterSpacing = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void MultiLine()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.5f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment + "\nLine 2");
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

}
