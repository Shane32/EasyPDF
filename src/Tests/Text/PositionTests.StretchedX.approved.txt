%PDF-1.4
%\xE2\xE3\xCF\xD3
2 0 obj
<</Length 918/Filter/FlateDecode>>stream
x\x9C\x8D\x97\xCBn\xDB@\x0CE\xF7\xFA
-\x93\x85Ur\xF4\x9C\xEE\x9A\xA2-\x10t\xD3\xD4\xBB"\x8B\xC2q\xD2\x14Q\xD2\xA4\x06\xFA\xFB\x1D\x91\xC3\x87\x8C\x18\x19\xE4azDr\xA8c\xF9^\xE9\xB9\xBA\xD8V\xEDP\x8F\xFDPoo\xAAO\xDB\xEA[\x855\xA4\x9F\xD6cH\xBFP\xEF\xE6
\xEA\xCB\xF4\xF7\xBB\x82\x06\xEB\x7F\xD5\x8F\xEBt\xFC\xA6\x02\xCA\xBB\xFA\x92\x83\x97\xBB\xD7\xD3B\x80\xA6O\xD1\x9C\xA3\xD0t\xB1\x0E8-/\xDD\xF2\x1E\x07z\xDDU\x01\xDB\xA6\xC7\xBC\x88\x96\x8AT\xBF\x1C_\xA2M^\xA5\xDC\x8Dt\xD8H\x0B\xEA\xCB\xCB\xB4]N\xCFC\xEC\xAA\xEF\x954\x9Cu\xB4\x07Z\xCDM\xE6J&Z\x96\x9F\xD3\x09mw\x0B\xA5w\x9F\xB1\xC6\xF4\xE6\xB6
\x02hI\x1C\x1B\x9C\xEA\xED\xEC!\xC0\x82\xF2\xC7\xD9\x87\x87\xFB\xBB\xC7y\xFFxx\x7F\xDE\xC3Y\xFDu\x7F{\xD8\x9Ec\x80\xB3\xA7?\xE7\xD7\xDBK\x86\xCD#\xA4\x16\x8A\x07\x1C\x9F\x10\xF2\xE9Q\xA0\x84\xF22\xBAt\xE4.\x02\x09{\xCA\xE3\xF4\xE5?u\xA1@!\xF12\xEF\xCF\xE92\x8BQr\x83\xA5P8Q%s\xA2Q\x8A@\x85\xD4/\x16\x93\xFA\x98\xDE\xED_\x8E9\xA5+U\xC6i'\x7F\x1D\x81\\H\xB0\xBA\x92@9I:r\x17\xE1\xD4\xB6\x8ES\x8B\xB9\x0B\x05\xCA\x89\x97y\xD3\xD68Q\x17\xE1\xE4\x07\x1B\x94\x13U\xE6\xEB\x09J9\xA5\xFARH\x17?\xFF\xEE\x1F\xEE\x1F\xF7\xC7\x98\xFAN\xA7\xE9\x07\x87\xA9\x9F\xF2\x09R\xA0\x98\xF22\xBAt\xE4.\x82\xA9G\x87\xA9\x8B\x02;zL\xBC\xCC\x9B\xA2a\xA2.\x82\xC9\x0F\xD6)&\xAAdL4J\x11\xA6\xB4Gl\xCBI=\x1D\x0EO\xF31\xA7$p2\xCE\xD89N\xE3\x90\xCF\x90\x02\xE5\x94\x97\xD1\xA5#w\x11NCt\x9C\x861w\xA1@9\xF12m\x9A\xD3e\x16\xE3\xE4\x07\x0B\xCA\x89*\x99\x13\x8D\xF2&'\\\xD4\xA1\xAB\xC7X\xAAQ\xFC\xAD;\xA5R\xD143z\x15\x8F"\xC2q\xA5\xE3\xD1\x84<:%\x8F&\xE5\xD3\xE8pM}\xEEB\x81\xE2\xE2e\xDA4\xA7\xCB,\x86\xCB\x0FfjN\x95\x8C+\x16\xC99B\xBA\xFASr[*T\x8C\xEBu\xA9Bp\xD2\x89^\xD4\x11E\xD59R`r\x00}\x09\xE6^\xAA\xEC\xB0\x92vPm\x87\xB5\xB8\x83Sw\xF0\xF2\x0E+}\xF7S\x82Sx0\x89\xE7\xB1\xDE\xA4\x17\xA7\x06\x035)GwJ\xC00\x98\x9Eb\xF0J\x8F\xADH=G\x06\xAF5\xB1\xD7\x12\xCC\xBD\x14^\xF0z\x8FA\x04\x9F#\x83\x17L\xF2\xA5D\xE7r\xF0VS\x9A\xECsy\x86\xD7\x16\x09?B\xD7\x8C\x916+T\xB5\x0C\xF0U]\xC3\xCEt\x16;\xEF\x00\xD8\x89\x05pd\xF8:3\x01-\xC1\xDCK\xF1u\xB8\xBA\xAD\x88\xFAQ\xAC\x9C \x1F\xC8\xDB;/\xE0^\x8Ao5e\xE7\xEE.\xCC\x0Fx\xAC7\xF1%w\x0F\xE9\xDAK\x9EP(tW\xF7w\xBFN\xDF\x8D&\xBF8xc\xC0A\x9C\x81#\xC37\x987h\x09\xE6^vS\xE6\xED\x01{\xF1\x07\x8E\x0C_o\x0E!%:\x97\xC3\xB7\x9A\xD2\\\x82\xCB3\xBE\xA1\xCC'\xA0\x09Kr_*|\x84\xEF\x84\xEEM\xE0n\x19\xBDM\xE0$>\xC1\x91\xC1\x9B\xCC)\xB4\x04s/\x857z\xB3\xC0Q\xDC\x82#\x837\x9A_H\x89\xCE\xE5\xEFk\xFD\x94\xE6\x19\\\x9E\xE1ME\xAE\xB1I_\x96@M\x8A\xC9\x9D\x94\xBD\xE8\x9F\x03V\x0F\x02\xA0O\x02\xB0~\x14\x00\xFF,\xE0=#:\xCF\x88+\xCF\x88\xEA\x19q\xED\x19\xD1yF\xF4\x9E\x11W\x9E\xE1\xA6\\be\x17\xDDc\x01\x14y\xC6\xD0\xC4@[\x15\x8A\x1E\xD3;\xD2\xBC\xFF\xEA\xC0\x03\x00
endstream
endobj
4 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]/Font<</F1 1 0 R>>>>/Contents 2 0 R/Parent 3 0 R>>
endobj
1 0 obj
<</Type/Font/Subtype/Type1/BaseFont/Helvetica/Encoding/WinAnsiEncoding>>
endobj
3 0 obj
<</Type/Pages/Count 1/Kids[4 0 R]/ITXT(3.3.2.0)>>
endobj
5 0 obj
<</Type/Catalog/Pages 3 0 R>>
endobj
6 0 obj
<<>>
endobj
xref
0 7
0000000000 65535 f 
0000001157 00000 n 
0000000015 00000 n 
0000001245 00000 n 
0000001000 00000 n 
0000001310 00000 n 
0000001355 00000 n 
trailer
<</Size 7/Root 5 0 R/Info 6 0 